[docs] Added info box for source code at beginning of doc (#20257)
MystenLabs/sui: docs/content/guides/developer/app-examples/tic-tac-toe.mdx

effort: small
---

:::info

You can view the [complete source code for this app example](https://github.com/MystenLabs/sui/tree/main/examples/tic-tac-toe) in the Sui repository.

:::

> This guide covers three different implementations of the game tic-tac-toe on Sui. The first example utilizes a centralized admin that owns the board object and marks it on the users’ behalf. The second example utilizes a shared object that both users can mutate. And the third example utilizes a multisig, where instead of sharing the game board, it's in a 1-of-2 multisig of both users’ accounts.

Author: Would it be better to add this when we bifurcate the two sections in the next PR?

> This guide covers three different implementations of the game tic-tac-toe on Sui. The first example utilizes a centralized admin that owns the board object and marks it on the users’ behalf. The second example utilizes a shared object that both users can mutate. And the third example utilizes a multisig, where instead of sharing the game board, it's in a 1-of-2 multisig of both users’ accounts.

Reviewer: It would.

  ```
</details>

:::checkpoint

- You have the latest version of Sui installed. If you run `sui --version` in your terminal or console, it responds with the currently installed version.
- You have a directory to place the files you create in. 
- You have created a `Move.toml` file in the `tic-tac-toe/move/` directory.

:::

## owned.move {#owned.move}

Create a new file in `tic-tac-toe/move/sources` titled `owned.move`. This file will contain the Move code for the game board in the centralized (and multisig) version of tic-tac-toe.

In this first example of tic-tac-toe, the `Game` object, including the game board, is controlled by a game admin.

{@inject: examples/tic-tac-toe/move/sources/owned.move#struct=Game noComments}

Ignore the `admin` field for now, as it is only relevant for the multisig approach.

Games are created with the `new` function:

{@inject: examples/tic-tac-toe/move/sources/owned.move#fun=new noComments}

Some things to notes: 
- `MARK__` is a constant that represents an empty cell on the game board. `MARK_X` and `MARK_O` represent the two players' markers.
- The first player is sent a `TurnCap` which gives them permission to take the next turn.

> - The `Game` object is created and returned by this function, it is up to its creator to send it to the game admin to own.

Reviewer: ```suggestion
- This function creates and returns the `Game` object, it is up to its creator to send it to the game admin to own.
```
Suggesting to put in active voice. Passive voice does put the "`Game` object" at the front of the sentence, which might be what you want to emphasize it. Dealer's choice.

> Because the players don’t own the game board, they cannot directly mutate it. Instead, they indicate their move by creating a `Mark` object with their intended placement and send it to the game object using transfer to object:

Reviewer: ```suggestion
Because the players don’t own the game board object, they cannot directly mutate it. Instead, they indicate their move by creating a `Mark` object with their intended placement and send it to the game object using transfer to object:
```
nittiest of nits

{@inject: examples/tic-tac-toe/move/sources/owned.move#struct=Mark noComments}

When playing the game, the admin operates a service that keeps track of marks using events. When a request is received (`send_mark`), the admin tries to place the marker on the board (`place_mark`). Each move requires two steps (thus two transactions): one from the player and one from the admin. This setup relies on the admin's service to keep the game moving.

{@inject: examples/tic-tac-toe/move/sources/owned.move#fun=send_mark,place_mark noComments}

> When a player sends a mark, a `Mark` object is created and is sent to the `Game` object. The admin then receives the mark and places it on the board. This is a use of dynamic object fields, where an object, `Game`, can hold other objects, `Mark`.

Reviewer: ```suggestion
When a player sends a mark, the contract creates a `Mark` object and sends it to the `Game` object. The admin then receives the mark and places it on the board. This is a use of dynamic object fields, where an object, `Game`, can hold other objects, `Mark`. To learn more about dynamic object fields, see [The Move Book](https://move-book.com/programmability/dynamic-object-fields.html).
```
Could just make "dynamic object fields" the link in the original last sentence, but adding an additional sentence lets us warn the users they'll be opening a new site.

Author: Should we be pointing readers to https://docs.sui.io/concepts/dynamic-fields or https://move-book.com/programmability/dynamic-object-fields.html ? Personally, I like the sui docs better than the move book's section on DoFs

Reviewer: Yeah, the docs version is going to go away soon. Let's leave the link off for now, though. I'll get with you later on how to integrate what you like in the one with the move book.

> To view the entire source code, see the [owned.move source file](https://github.com/MystenLabs/sui/blob/main/examples/tic-tac-toe/move/sources/owned.move). You can find the rest of the logic, including how to check for a winner, as well as deleting the gameboard after the game concludes there.

Reviewer: ```suggestion
To view the entire source code, see the [owned.move source file](https://github.com/MystenLabs/sui/blob/main/examples/tic-tac-toe/move/sources/owned.move). You can find the rest of the logic, including how to check for a winner, as well as deleting the game board after the game concludes there.
```

}
```

> `useTransactions.ts` also contains functions to place, send, and receive marks, end the game, and burn ended games. These functions all return a `Transaction` object which is used in the React frontend to execute the transaction with the appropriate signer.

Reviewer: ```suggestion
`useTransactions.ts` also contains functions to place, send, and receive marks, end the game, and burn completed games. These functions all return a `Transaction` object, which is used in the React frontend to execute the transaction with the appropriate signer.
```
